xbmc/addons/AddonRepos.cpp

```diff
@@ -348,6 +368,64 @@ void CAddonRepos::GetLatestAddonVersions(std::vector<std::shared_ptr<IAddon>>& a
   }
 }
 
+void CAddonRepos::GetLatestAddonVersionsFromAllRepos(
+    std::vector<std::shared_ptr<IAddon>>& addonList) const
+{
+  const AddonRepoUpdateMode updateMode =
+      CAddonSystemSettings::GetInstance().GetAddonRepoUpdateMode();
+
+  addonList.clear();
+
+  // first we insert all official addon versions into the resulting vector
+
+  std::transform(m_latestOfficialVersions.begin(), m_latestOfficialVersions.end(),
+                 back_inserter(addonList),
+                 [](const std::pair<std::string, std::shared_ptr<IAddon>>& officialVersion) {
+                   return officialVersion.second;
+                 });
+
+  // then we insert latest version per addon and repository if they don't exist in the official map
+  // or installation from ANY_REPOSITORY is allowed and the private version is higher
+
+  for (const auto& repo : m_addonsByRepoMap)
+  {
+    // content of official repos is stored in m_addonsByRepoMap too
+    // so we need to filter them out
+
+    if (std::none_of(
+            officialRepoInfos.begin(), officialRepoInfos.end(),
+            [&](const ADDON::RepoInfo& officialRepo) { return repo.first == officialRepo.m_repoId; }))
+    {
+      std::string rangeId = repo.second.begin()->first;
+      std::shared_ptr<IAddon> latestAddon;
+
+      do
+      {
+        const auto& addonRange = repo.second.equal_range(rangeId);
+
+        const auto& officialVersion = m_latestOfficialVersions.find(rangeId);
+        if (officialVersion == m_latestOfficialVersions.end() ||
+            updateMode == AddonRepoUpdateMode::ANY_REPOSITORY)
+        {
+          // find latest version of addon and emplace
+
+          latestAddon.reset();
+          for (auto addon = addonRange.first; addon != addonRange.second; ++addon)
+          {
+            if (!latestAddon || addon->second->Version() > latestAddon->Version())
+            {
+              latestAddon = addon->second;
+            }
+          }
+          addonList.emplace_back(latestAddon);
+
+          rangeId = addonRange.second == repo.second.end() ? "" : addonRange.second->first;
+        }
+      } while (!rangeId.empty());
```

Comment:
About here I'm not sure if in all cases good as `rangeId` becomes changed within `if`.

Not more safe to use `} while (addonRange.second == repo.second.end());` here?

Reply:
upsi... this is wrong. `rangeId` needs to be set outside `if`